clang/lib/Sema/TreeTransform.h

```diff
@@ -7668,6 +7668,39 @@ QualType TreeTransform<Derived>::TransformAttributedType(TypeLocBuilder &TLB,
       }
     }
 
+    /* TO_UPSTREAM(BoundsSafety) ON*/
+    // This is primarily here for the `TemplateInstantiator` subclass
+    // where we want to check that after template instantiation the
+    // bounds-safety attributes are being applied to pointer types.
+    if (SemaRef.LangOpts.isBoundsSafetyAttributeOnlyMode()) {
+      const char *AttrName = nullptr;
+      switch (TL.getAttrKind()) {
+      case attr::PtrSingle:
+        AttrName = "__single";
+        LLVM_FALLTHROUGH;
+      case attr::PtrUnsafeIndexable: {
+        AttrName = AttrName ? AttrName : "__unsafe_indexable";
+        QualType ShouldBePtrTy =
+            modifiedType.getDesugaredType(SemaRef.getASTContext());
+        if (!ShouldBePtrTy->isPointerType()) {
+          const auto *Attr = TL.getAttr();
+          // Use the attribute spelling if it's available.
+          // FIXME: When would it not be available?
```

Comment:
There may be other cases, but if the code was ever modified so that `-fbounds-safety` could reach here, I think implicitly single pointers would probably miss spelling